extension/js/common/message-renderer.ts

```diff
@@ -0,0 +1,782 @@
+      if (fallbackToPlainText) {
+        renderModule.renderAsRegularContent(Str.with(encryptedData));
+      } else {
+        renderModule.renderErr(Lang.pgpBlock.badFormat + '\n\n' + result.error.message, Str.with(encryptedData));
+      }
+    } else if (result.longids.needPassphrase.length) {
+      renderModule.renderPassphraseNeeded(result.longids.needPassphrase);
+      return { needPassphrase: result.longids.needPassphrase };
+    } else {
+      if (!result.longids.chosen && !(await KeyStore.get(this.acctEmail)).length) {
+        renderModule.renderErr(Lang.pgpBlock.notProperlySetUp + MessageRenderer.btnHtml('FlowCrypt settings', 'green settings'), undefined);
+      } else if (result.error.type === DecryptErrTypes.keyMismatch) {
+        await MessageRenderer.handlePrivateKeyMismatch(
+          renderModule,
+          kisWithPp.map(ki => ki.public),
+          encryptedData,
+          isPwdMsgBasedOnMsgSnippet === true
+        );
+      } else if (result.error.type === DecryptErrTypes.wrongPwd || result.error.type === DecryptErrTypes.usePassword) {
+        renderModule.renderErr(Lang.pgpBlock.pwdMsgAskSenderUsePubkey, undefined);
+      } else if (result.error.type === DecryptErrTypes.noMdc) {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        renderModule.renderErr(result.error.message, result.content!.toUtfStr()); // missing mdc - only render the result after user confirmation
+      } else if (result.error) {
+        renderModule.renderErr(`${Lang.pgpBlock.cantOpen}\n\n<em>${result.error.type}: ${result.error.message}</em>`, Str.with(encryptedData));
+      } else {
+        // should generally not happen
+        renderModule.renderErr(
+          Lang.pgpBlock.cantOpen +
+            Lang.general.writeMeToFixIt(await isCustomerUrlFesUsed(this.acctEmail)) +
+            '\n\nDiagnostic info: "' +
+            JSON.stringify(result) +
+            '"',
+          Str.with(encryptedData)
+        );
+      }
+    }
+    return {};
+  };
+
+  private setMsgBodyAndStartProcessing = async (
+    loaderContext: LoaderContextInterface,
+    type: string, // for diagnostics
+    printMailInfo: PrintMailInfo | undefined,
+    senderEmail: string | undefined,
+    cb: (renderModule: RenderInterface, frameId: string) => Promise<{ publicKeys?: string[] }>
+  ) => {
+    const { frameId, frameXssSafe } = loaderContext.factory.embeddedRenderMsg(type);
+    loaderContext.setMsgBody(frameXssSafe, 'set');
+    await this.relayAndStartProcessing(this.relayManager, loaderContext, loaderContext.factory, frameId, printMailInfo, senderEmail, cb);
+  };
+
+  private processMessageWithDetachedSignature = async (msgId: string, renderModule: RenderInterface, senderEmail: string | undefined) => {
+    try {
+      renderModule.renderText('Loading signed message...');
+      const raw = await this.msgGetRaw(msgId); // todo: can we try to restore the content attachment from 'full'?
```

Comment:
It's not easy to implement the previous "Retry" feature (when failing to get 'raw' message for PGP/MIME signed message)
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/23335944/780e8827-0c80-48d6-a31a-3192f32d2cef)

Should I implement or can we replace with a simpler UI?
@tomholub

Reply:
Or we can create pgp_block after loading 'raw' first and display error above the native gmail attachments like this
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/23335944/cca1a317-b4a9-43f6-93cf-9c73a07a2f0b)

Reply:
Second option is ok, but while loading, there should be progress visible. It may take a very long time for a message to load, someone may otherwise thing the extension is broken.

Reply:
> But while loading, there should be progress visible. It may take a very long time for a message to load, someone may otherwise think the extension is broken.

By means of a spinner?
I see it is currently broken. I posted #5196

Reply:
Some sort of indication that a message is loading, be it a spinner, percentage, or just a "Loading message..." text at a minimum.

Reply:
Implemented like this:
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/23335944/c503044b-5634-4756-aae6-197520e53247)
This status is rendered for all attachments when they are being analyzed. The status remains unchanged when 'raw' message is fetched to process this signature.

Then, if network failure happens at this stage, the displayed status would be `Categorize: net err` (I wonder why such strange text was already in place)